common/transport/amqp/devdoc/sender_link_requirements.md

```diff
@@ -0,0 +1,78 @@
+
+## Overview
+
+The `SenderLink` class is internal to the SDK and shall be considered private. It shall not be used directly by clients of the SDK, who should instead rely on the higher-level constructs (`Client` classes provided by `azure-iot-device` and `azure-iothub`).
+
+The `SenderLink` class implements a state machine that manages the underlying `amqp10` link object used to send messages to IoT Hub. It can be attached and detached manually, and will try to attach automatically if not already attached when calling `send`.
+
+## Example usage
+
+```typescript
+import * as amqp10 from 'amqp10';
+import { AmqpMessage } from './amqp_message';
+
+const linkAddress = 'exampleAddress';
+const amqp10Client = new amqp10.AmqpClient(null);
+
+const senderLink = new SenderLink(linkAddress, null, amqp10Client);
+senderLink.on('errorReceived', (err) => {
+  console.error(err.toString());
+});
+
+senderLink.send(new AmqpMessage(''), linkAddress, (err) => {
+  if (err) {
+    console.error(err.toString());
+  } else {
+    console.log('message successfully sent');
+  }
+});
+```
+
+## Public Interface
+
+### constructor(linkAddress: string, linkOptions: any, amqp10Client: amqp10.AmqpClient)
+
+**SRS_NODE_AMQP_SENDER_LINK_16_002: [** The `SenderLink` class shall inherit from `EventEmitter`. **]**
+
+**SRS_NODE_AMQP_SENDER_LINK_16_003: [** The `SenderLink` class shall implement the `AmqpLink` interface. **]**
+
+### attach(callback: (err?: Error) => void): void
+
+**SRS_NODE_AMQP_SENDER_LINK_16_004: [** The `attach` method shall use the stored instance of the `amqp10.AmqpClient` object to attach a new link object with the `linkAddress` and `linkOptions` provided when creating the `SenderLink` instance. **]**
+
+**SRS_NODE_AMQP_SENDER_LINK_16_022: [** The `attach` method shall call the `callback` if the link was successfully attached. **]**
+
+**SRS_NODE_AMQP_SENDER_LINK_16_006: [** The `SenderLink` object should subscribe to the `detached` event of the newly created `amqp10` link object. **]**
+
+**SRS_NODE_AMQP_SENDER_LINK_16_007: [** The `SenderLink` object should subscribe to the `errorReceived` event of the newly created `amqp10` link object. **]**
+
+**SRS_NODE_AMQP_SENDER_LINK_16_008: [** If the `amqp10.AmqpClient` fails to create the link the `callback` function shall be called with this error object. **]**
+
+### detach(): void
+
+**SRS_NODE_AMQP_SENDER_LINK_16_009: [** The `detach` method shall detach the link created by the `amqp10.AmqpClient` underlying object. **]**
+
+### send(message: AmqpMessage, callback: (err?: Error, result?: results.MessageEnqueued) => void): void
+
+**SRS_NODE_AMQP_SENDER_LINK_16_010: [** The `send` method shall use the link created by the underlying `amqp10.AmqpClient` to send the specified `message` to the IoT hub. **]**
+
+**SRS_NODE_AMQP_SENDER_LINK_16_011: [** If the state machine is not in the `attached` state, the `SenderLink` object shall attach the link first and then send the message. **]**
```

Comment:
>If the state machine is not in the `attached` state, the `SenderLink` object shall attach the link first and then send the message. [](start = 40, length = 131)

This is inconsistent with the complete, reject, and abandon functions on the receiver link object.  Why does this attach the link, but those functions don't?

Reply:
because the message lock is lost if the link is detached and therefore complete/reject/abandon cannot possibly succeed, whereas send can.

---
In reply to: [139494826](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139494826) [](ancestors = 139494826)